### DISCUSSION TITLE
Using NSubstitute within AssertionScope

### DISCUSSION THREAD

**Discussion by siewers** [fluentassertions/fluentassertions#2800] [Q&A]
Hi!

I just recently noticed that any non-test framework exception exits the assertion scope prematurely, for example:

```cs
public class Tests
{
    [Fact]
    public void AssertionsTest()
    {
        int value1 = 1;
        var service = Substitute.For<IService>();
        int value2 = 2;

        using (new AssertionScope())
        {
            service.Received(1).Method();
            value1.Should().Be(1);
            value2.Should().NotBe(2);
        }
    }

    public interface IService
    {
        void Method();
    }
}
```

In the above example, if the call using NSubstitute fails, the subsequent assertions will not be executed, and the test output will only contain the one exception produced by NSubstitute:
```
NSubstitute.Exceptions.ReceivedCallsException: Expected to receive exactly 1 call matching:

NSubstitute.Exceptions.ReceivedCallsException
Expected to receive exactly 1 call matching:
	Method()
Actually received no matching calls.
```

If I rearrange the assertions, it somehow ignores the exception thrown by NSubstitute:
```cs
value1.Should().Be(1);
value2.Should().NotBe(2);
service.Received(1).Method();
```
Result:
```
Xunit.Sdk.XunitException
Did not expect value2 to be 2.
```

The behavior is a bit difficult to understand.

I tried working on something that could bridge the gap between NSubstitute and FluentAssertions, but the best I could come up with is not that nice and forces you to remember how to write the assertions:

```cs
public static void ShouldHaveReceived<T>(this T substitute, int requiredNumberOfCalls, Action<T> action)
    where T : class
{
    using var scope = new AssertionScope();

    try
    {
        action(substitute.Received(Quantity.Exactly(requiredNumberOfCalls)));
    }
    catch (ReceivedCallsException exception)
    {
        scope.FailWith(exception.Message);
    }
}
```

This allows me to write this:
```
using (new AssertionScope())
{
    value1.Should().Be(1);
    value2.Should().NotBe(2);
    service.ShouldHaveReceived(1, _ => service.Method());
}
```
Giving me this output:
```
Xunit.Sdk.XunitException
Did not expect value2 to be 2.
Expected to receive exactly 1 call matching:
	Method()
Actually received no matching calls.
```

Is there any way to handle specific exceptions produced by other testing libraries?
In this case, I'd like to handle the `NSubstitute.Exceptions.ReceivedCallsException` exception type in the `AssertionScope` so that all assertion failures are collected correctly.

I was hoping for some global setting where specific exceptions could be configured to be handled in assertion scopes, but I couldn't find anything.

I imagined something like this:
```cs
GlobalAssertionScopeConfiguration.DefineAssertionExceptionHandler<ReceivedCallsException>(
    ex => /* Format exception message if needed */
);
```

As a final note, it should be mentioned that I might be thinking about this the wrong way and that I shouldn't be mixing testing libraries like this. However, I believe this is a widespread practice, and NSubstitute + FluentAssertions is a powerful combination.

**Comment by dennisdoomen** [comment]
Unfortunately that won't be possible. The `AssertionScope` is not some magic exception catching thing. All the FA APIs explicitly interact with it through `AssertionChain.FailWith`.

**Reply by siewers** [reply]
I understand that, but do you have any suggestions on how to handle this case in a better way?

**Reply by dennisdoomen** [reply]
Maybe a `try...catch`? For the rest there's no relationship between those two frameworks.

**Reply by siewers** [reply]
Okay. Can you explain why the NSubstitute exception disappears when it's the last assertion done in the scope? Something in the scope must be doing something with the exception since it's not thrown?

**Reply by dennisdoomen** [reply]
That's because the closing curly brace triggers a call to `IDisposable.Dispose` on the `AssertionScope`. Since that one throws because of the failing assertions, the other exception thrown gets ignored. That's a limitation of .NET.
